test_programs/compile_success_no_bug/noirc_frontend_tests_traits_implicit_generics_unification/src/main.nr

```diff
@@ -0,0 +1,38 @@
+
+    trait TraitWithAssociatedConstant {
+        let N: u32;
+
+        fn foo(_: Self) -> bool {
+            true
+        }
+    }
+
+    struct Foo {}
```

Comment:
Is this the right repro? This passes on master

Reply:
Testing against the original code in https://github.com/noir-lang/noir/issues/8252 I still get the same error with this PR

Reply:
Gah, I copied the second snippet in the issue, which is very similar to the first one and works fine. I'll update the test.

Reply:
Huh, it doesn't even reproduce if I don't use the `Eq` trait from the stdlib and `assert_eq`.

Reply:
This is interesting. In the original snippet we have:

```noir
    assert_eq(wrapper, wrapper);
```

That fails to compile. If we change it to this:

```noir
    assert(wrapper == wrapper);
```

it still fails to compile.

But if we change it to this:

```noir
    assert(wrapper.eq(wrapper));
```

then it compiles, so I'm starting to suspect it has something to do with operator overloading.

Still need to update the test, I'll do it now.

Reply:
Done!